beacon-chain/slasher/detect_attestations.go

```diff
@@ -119,6 +159,79 @@ func (s *Service) determineChunksToUpdateForValidators(
 	return
 }
 
+// Checks if an incoming attestation is slashable based on the validator chunk it
+// corresponds to. If a slashable offense is found, we return it to the caller.
+// If not, then update every single chunk the attestation covers, starting from its
+// source epoch up to its target.
+func (s *Service) applyAttestationForValidator(
+	ctx context.Context,
+	opts *chunkUpdateOptions,
+	chunksByChunkIdx map[uint64]Chunker,
+	attestation *slashertypes.CompactAttestation,
+) (*slashertypes.Slashing, error) {
+	sourceEpoch := types.Epoch(attestation.Source)
+	targetEpoch := types.Epoch(attestation.Target)
+	chunkIdx := s.params.chunkIndex(sourceEpoch)
+	chunk, ok := chunksByChunkIdx[chunkIdx]
+	if !ok {
+		return nil, nil
+	}
+
+	// Check slashable, if so, return the slashing.
+	slashing, err := chunk.CheckSlashable(
+		ctx,
+		s.serviceCfg.Database,
+		opts.validatorIndex,
+		attestation,
+	)
+	if err != nil {
+		return nil, err
+	}
+	if slashing != nil && slashing.Kind != slashertypes.NotSlashable {
+		return slashing, nil
+	}
+
+	// Get the first start epoch for the chunk. If it does not exist or
+	// is not possible based on the input arguments, do not continue with the update.
+	startEpoch, exists := chunk.StartEpoch(sourceEpoch, opts.currentEpoch)
+	if !exists {
+		return nil, nil
+	}
+
+	// Given a single attestation could span across multiple chunks
+	// for a validator min or max span, we attempt to update the current chunk
+	// for the source epoch of the attestation. If the update function tells
+	// us we need to proceed to the next chunk, we continue by determining
+	// the start epoch of the next chunk. We exit once no longer need to
+	// keep updating chunks.
+	for {
```

Comment:
@rkapka this is the update routine we talked about. The reason that Update returns a keep going is that this keep going variable's condition differs whether it is a min span chunk or max span chunk. To make it its own pure function would lead us to have to duplicate this entire function, making our code less DRY. Curious what you think, as IMO, this seems fairly clean as a for loop

Reply:
@rauljordan So you ultimately decided not to make `Update` a service method, but keep a chunk method? I thought the idea was to move `Update` to the service, so that the chunk does not update itself and does not convey information about other chunks.

Reply:
I guess that's all right if it's deliberate, we also have `chunk.NextChunkStartEpoch` which again results in one chuck knowing about another chunk.

The `for` loop is very clean and readable and the code works. Let's be practical and keep it this way. We can always refactor it if this design causes issues in the future.